src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -360,12 +375,39 @@ private bool TryHandleAsKeySegment(string segmentText)
             KeySegment previousKeySegment = this.FindPreviousKeySegment();
 
             KeySegment keySegment;
+
+            bool trybindingEscapeFunction = false;
+
+            if (segmentText.Length > 0 && segmentText[segmentText.Length - 1] == ':' && previous.EdmType != null)
+            {
+                IEdmFunction function = configuration.Model.FindBoundOperations(previous.EdmType).OfType<IEdmFunction>().FirstOrDefault(f => IsUrlEscapeFunction(configuration.Model, f));
+                if (function != null)
+                {
+                    trybindingEscapeFunction = true;
+                    segmentText = segmentText.Substring(0, segmentText.Length - 1);
+
+                    // If nothing to bind for key then bind the escape function directly. 
+                    if (segmentText.Length == 0)
+                    {
+                        return this.TryBindEscapeFunction();

```

Comment:
You have already done a somewhat expensive operation to figure out what the escape function is. You shouldn't have to redo that work when you call TryBindEscapeFunction. Can you split TryBindEscapeFunction into two pieces, the first the determines the function and the second that takes that information and does the actual binding, and call the second function from here and line 403?

Reply:
Instead of splitting it into two functions I decided to have this take in an optional function.

---
In reply to: [366150654](https://github.com/OData/odata.net/pull/1621#discussion_r366150654) [](ancestors = 366150654)

Reply:
See my comments in iteration 22. I still think we can do a little refactoring to make the code cleaner and more performant.

---
In reply to: [366662603](https://github.com/OData/odata.net/pull/1621#discussion_r366662603) [](ancestors = 366662603,366150654)